src/OpenTelemetry/Logs/LogRecord.cs

```diff
@@ -134,21 +141,28 @@ public void ForEachScope<TState>(Action<LogRecordScope, TState> callback, TState
         }
 
         /// <summary>
-        /// Buffers the scopes attached to the log into a list so that they can
+        /// Buffers the states &amp; scopes attached to the log so that they can
         /// be safely processed after the log message lifecycle has ended.
         /// </summary>
-        internal void BufferLogScopes()
+        internal void Buffer()
         {
-            if (this.ScopeProvider == null || this.bufferedScopes != null)
+            if (this.stateValues != null && this.bufferedStateValues == null)
             {
-                return;
+                // Note: We copy the state values to capture anything deferred.
+                // See:
+                // https://github.com/open-telemetry/opentelemetry-dotnet/issues/2905
+
+                this.bufferedStateValues = new(this.stateValues);
```

Comment:
does this copy the whole collection to bufferedStateValues? if not, we'll still have the same issue, when we eventually iteration on bufferedStateValues?

Reply:
It definitely works because I started with a failing test and then I made it pass. BUT... there are two branches in [`List<T>`](https://github.com/dotnet/runtime/blob/7984b32774916c98ab7c85c244c9e40581e4cdf5/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L66-L90) I was expecting us to hit the first one which I thought would be faster. Looking at it right now, we are probably falling into the second one which just does an enumeration. Because `IReadOnlyList<T>` does not implement `ICollection<T>`. Oh .NET collections! That enumeration is probably costing us some allocation we can avoid. Let me perf test it a bit and I'll reply back.

Reply:
@cijothomas OK I did some perf testing. Perf depends on the concrete type of the state. When there is an `ICollection<T>` the `CopyTo` directly over the array gives a nice boost. When there is only `IReadOnlyList` we can avoid an allocation using a `for` loop instead of using the enumerator. For other states, we already make a copy so there isn't any need to buffer at all. I added some type checking to optimize for those cases.

Let me know what you think about that.

|              Method | NumberOfItemsInState |      Mean |     Error |    StdDev |  Gen 0 |  Gen 1 | Allocated |
|-------------------- |--------------------- |----------:|----------:|----------:|-------:|-------:|----------:|
|     BufferListState |                    0 |  38.79 ns |  0.480 ns |  0.401 ns | 0.0108 |      - |     136 B |
| BufferReadOnlyState |                    0 |  37.15 ns |  0.324 ns |  0.253 ns | 0.0108 |      - |     136 B |
|   BufferCopiedState |                    0 |  39.59 ns |  0.565 ns |  0.472 ns | 0.0108 |      - |     136 B |
|     BufferListState |                  100 | 142.18 ns |  1.753 ns |  1.369 ns | 0.1402 | 0.0007 |   1,760 B |
| BufferReadOnlyState |                  100 | 648.79 ns | 11.549 ns | 10.803 ns | 0.1402 |      - |   1,760 B |
|   BufferCopiedState |                  100 |  39.69 ns |  0.807 ns |  0.829 ns | 0.0108 |      - |     136 B |